combine/combine.go

```diff
@@ -0,0 +1,226 @@
+		}
+
+		secret, err := tblsv2.RecoverSecret(shares, uint(len(lock.Operators)), uint(lock.Threshold))
+		if err != nil {
+			return errors.Wrap(err, "cannot recover shares", z.Int("validator_number", idx))
+		}
+
+		// require that the generated secret pubkey matches what's in the lockfile for the idx validator
+		val := lock.Validators[idx]
+
+		valPk, err := val.PublicKey()
+		if err != nil {
+			return errors.Wrap(err, "public key for validator from lockfile", z.Int("validator_number", idx))
+		}
+
+		genPubkey, err := tblsv2.SecretToPublicKey(secret)
+		if err != nil {
+			return errors.Wrap(err, "public key for validator from generated secret", z.Int("validator_number", idx))
+		}
+
+		if valPk != genPubkey {
+			return errors.New("generated and lockfile public key for validator DO NOT match", z.Int("validator_number", idx))
+		}
+
+		outPath := filepath.Join(inputDir, val.PublicKeyHex())
+		if err := os.Mkdir(outPath, 0o755); err != nil {
+			return errors.Wrap(err, "output directory creation", z.Int("validator_number", idx))
+		}
+
+		ksPath := filepath.Join(outPath, "keystore-0.json")
+		_, err = os.Stat(ksPath)
+		if err == nil && !force {
+			return errors.New("refusing to overwrite existing private key", z.Int("validator_number", idx), z.Str("path", ksPath))
+		}
+
+		if err := keystore.StoreKeys([]tblsv2.PrivateKey{secret}, outPath); err != nil {
+			return errors.Wrap(err, "cannot store keystore", z.Int("validator_number", idx))
+		}
+	}
+
+	return nil
+}
+
+func secretsToShares(lock cluster.Lock, secrets []tblsv2.PrivateKey) (map[int]tblsv2.PrivateKey, error) {
+	n := len(lock.Operators)
+
+	resp := make(map[int]tblsv2.PrivateKey)
+	for idx, secret := range secrets {
+		pubkey, err := tblsv2.SecretToPublicKey(secret)
+		if err != nil {
+			return nil, errors.Wrap(err, "pubkey from share")
+		}
+
+		var found bool
+		for _, val := range lock.Validators {
+			for i := 0; i < n; i++ {
+				pubShare, err := val.PublicShare(i)
+				if err != nil {
+					return nil, errors.Wrap(err, "pubshare from lock")
+				}
+
+				if !bytes.Equal(pubkey[:], pubShare[:]) {
+					continue
+				}
+
+				resp[idx+1] = secret
+				found = true
+
+				break
+			}
+
+			if found {
+				break
+			}
+		}
+
+		if !found {
+			return nil, errors.New("share not found in lock")
+		}
+	}
+
+	return resp, nil
+}
+
+// loadLockfile loads a lockfile from one of the charon directories contained in dir.
+// It checks that all the directories containing a validator_keys subdirectory contain the same cluster_lock.json file.
+// It returns the cluster.Lock read from the lock file, and a list of directories that possibly contains keys.
+func loadLockfile(dir string) (cluster.Lock, []string, error) {
+	root, err := os.ReadDir(dir)
+	if err != nil {
+		return cluster.Lock{}, nil, errors.Wrap(err, "can't read directory")
+	}
+
+	lfFound := false
+	lastLockfileHash := [32]byte{}
+	lfContent := []byte{}
+	possibleValKeysDir := []string{}
```

Comment:
nit: use `var` style when declaring but not initialising variables: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_use_a_consistent_declaration_style

Reply:
```
var (
  lfFound bool
  lastLockFileHash [32]byte
  ...
)
```

Reply:
Yup, will refactor.

Old habits die hard.